nexus/db-model/src/deployment.rs

```diff
@@ -256,82 +266,442 @@ impl BpOmicronZone {
         blueprint_id: Uuid,
         sled_id: SledUuid,
         blueprint_zone: &BlueprintZoneConfig,
-    ) -> Result<Self, anyhow::Error> {
+    ) -> anyhow::Result<Self> {
         let external_ip_id = blueprint_zone
             .zone_type
             .external_networking()
-            .map(|(ip, _)| ip.id());
-        let zone = OmicronZone::new(
-            sled_id,
-            blueprint_zone.id.into_untyped_uuid(),
-            blueprint_zone.underlay_address,
-            blueprint_zone.filesystem_pool.as_ref().map(|pool| pool.id()),
-            &blueprint_zone.zone_type.clone().into(),
-            external_ip_id,
-        )?;
-        Ok(Self {
+            .map(|(ip, _)| ip.id().into());
+
+        // Create a dummy record to start, then fill in the rest
+        let mut bp_omicron_zone = BpOmicronZone {
+            // Fill in the known fields that don't require inspecting
+            // `blueprint_zone.zone_type`
             blueprint_id,
-            sled_id: zone.sled_id.into(),
-            id: zone.id,
-            underlay_address: zone.underlay_address,
-            zone_type: zone.zone_type,
-            primary_service_ip: zone.primary_service_ip,
-            primary_service_port: zone.primary_service_port,
-            second_service_ip: zone.second_service_ip,
-            second_service_port: zone.second_service_port,
-            dataset_zpool_name: zone.dataset_zpool_name,
-            bp_nic_id: zone.nic_id,
-            dns_gz_address: zone.dns_gz_address,
-            dns_gz_address_index: zone.dns_gz_address_index,
-            ntp_ntp_servers: zone.ntp_ntp_servers,
-            ntp_dns_servers: zone.ntp_dns_servers,
-            ntp_domain: zone.ntp_domain,
-            nexus_external_tls: zone.nexus_external_tls,
-            nexus_external_dns_servers: zone.nexus_external_dns_servers,
-            snat_ip: zone.snat_ip,
-            snat_first_port: zone.snat_first_port,
-            snat_last_port: zone.snat_last_port,
-            disposition: to_db_bp_zone_disposition(blueprint_zone.disposition),
-            external_ip_id: zone.external_ip_id.map(From::from),
+            sled_id: sled_id.into(),
+            id: blueprint_zone.id.into_untyped_uuid(),
+            underlay_address: blueprint_zone.underlay_address.into(),
+            external_ip_id,
             filesystem_pool: blueprint_zone
                 .filesystem_pool
                 .as_ref()
                 .map(|pool| pool.id().into()),
-        })
+
+            // Set the remainder of the fields to a default
+            disposition: DbBpZoneDisposition::InService,
+            zone_type: ZoneType::BoundaryNtp,
```

Comment:
I think this could be

```suggestion
            zone_type: blueprint_zone.zone_type.kind().into(),
```

to avoid having to set it by hand in each case below.